pkg/dockerregistry/server/writelimiter/blobstore.go

```diff
@@ -0,0 +1,107 @@
+package writelimiter
+
+import (
+	"runtime"
+	"sync/atomic"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+
+	"github.com/openshift/origin/pkg/dockerregistry/server/types"
+)
+
+type blobWriter struct {
+	distribution.BlobWriter
+	release func()
+}
+
+var _ distribution.BlobWriter = blobWriter{}
+
+func (bw blobWriter) Close() error {
+	bw.release()
+	return bw.BlobWriter.Close()
+}
+
+type blobStore struct {
+	distribution.BlobStore
+	sem chan struct{}
+}
+
+var _ distribution.BlobStore = blobStore{}
+
+func (bs blobStore) acquire() func() {
+	// LOCK_BEGIN++
+	sem := bs.sem
+	// LOCK_WAIT_BEGIN++
+	sem <- struct{}{}
+	// LOCK_WAIT_END++
+	done := int32(0)
+	return func() {
+		if atomic.SwapInt32(&done, 1) == 0 {
+			<-sem
+			sem = nil
```

Comment:
The meaning behind it is to release the channel from the closure. Not very important, but it's not a left-over.